### PR TITLE
Introduce float8 training

### PR DESCRIPTION
**Pull request by james77777778**
This PR introduces float8 training, especially for TF and JAX.

### A New Attribute `overwrite_with_gradient` Has Been Added to `Variable`

This attribute will be utilized in `Optimizer` if set to `True`. The optimizer will directly overwrite the variable by the computed gradient in `BaseOptimizer._overwrite_variables_directly_with_gradients` of `BaseOptimizer.apply`.

With this new behavior, we can correctly update float8 parameters such as `scale` and `amax_history`.

### A New Quantization Mode `float8` Has Been Added

To `DTypePolicy`:
```python
# This should be only invoked by quantizable layers.
original_policy = dtype_policy().name
policy = QuantizedDTypePolicy(f"float8_from_{original_policy}")
```

To `Dense` and `EinsumDense`:
```python
# Dense
model = models.Sequential([layers.Dense(units=16)])
model.quantize("float8")
model.compile(optimizer="sgd", loss="mse")
model.fit(x, y, epochs=2)

# EinsumDense
model = models.Sequential([layers.EinsumDense("ab,bcd->acd", (8, 32))])
model.quantize("float8")
model.compile(optimizer="sgd", loss="mse")
model.fit(x, y, epochs=2)
```

The compiled graph in TF should contains the expected op `__cublas$lt$matmul$f8`:
fp8

<details>
<summary>generate_hlo_dot.py</summary>

```python
import tensorflow as tf
from keras import backend
from keras import config
from keras import layers
from keras import ops

keras.config.set_dtype_policy("mixed_bfloat16")

layer = layers.EinsumDense("abcde,aebf->adbcf", output_shape=[16, 16, 16, 16])
layer.build([32, 16, 16, 16, 16])
x = ops.ones([32, 16, 16, 16, 16])
layer.quantize("float8")

def fn(x):
    return layer(x)

compiled_fn = tf.function(fn, jit_compile=True)
y = compiled_fn(x)
ir = compiled_fn.experimental_get_compiler_ir(x)(stage="optimized_hlo_dot")
with open("ir_tf.dot", "w") as f:
    f.write(ir)

```

</details>

## Results

- RTX 4070 12GB, CUDA12.3, CUDNN8.9
- GPT2 Base: batch_size=64, frozen embeddings
- GPT2 Medium: batch_size=24, frozen embeddings

| Backend | Model       | Recipe | Training Speed  | Peak Mem. Usage | Val. Acc. |
|---------|-------------|-----|-----------------|-----------------|-----------|
| TF      | GPT2 Base   | `mixed_bfloat16`   | 274~275 ms/step | 6.234GB         | 0.4997    |
| TF      | GPT2 Base   | `float8`   | 269~272 ms/step | 5.909GB         | 0.4971    |
| TF      | GPT2 Medium | `mixed_bfloat16`   | 252~254 ms/step | 7.338GB         | 0.5535    |
| TF      | GPT2 Medium | `float8`   | 268~271 ms/step | 7.002GB         | 0.5513    |
| JAX     | GPT2 Base   | `mixed_bfloat16`   | 207~208 ms/step | X               | 0.4991    |
| JAX     | GPT2 Base   | `float8`   | 200~202 ms/step | X               | 0.4977    |
| JAX     | GPT2 Medium | `mixed_bfloat16`   | 236~238 ms/step | X               | 0.5534    |
| JAX     | GPT2 Medium | `float8`   | 232~234 ms/step | X               | 0.5521    |

Standalone script:

<details>
<summary>fp8.py</summary>

```python
"""
Dataset:
https://huggingface.co/datasets/databricks/databricks-dolly-15k/blob/main/databricks-dolly-15k.jsonl
"""

import argparse
import json

import kagglehub
import keras_nlp

import keras


def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--model",
        default="gpt2_base_en",
        choices=["gpt2_base_en", "gpt2_medium_en", "gemma_2b_en"],
        help="Which model to demonstrate",
    )
    parser.add_argument(
        "--fp8",
        action="store_true",
        help="Whether to use float8 technique",
    )
    parser.add_argument("-b", "--batch-size", default=32, type=int)
    args = parser.parse_args()
    return args


def get_optimizer_and_loss():
    optimizer = keras.optimizers.AdamW(
        learning_rate=5e-5,
        weight_decay=0.01,
        epsilon=1e-6,
        global_clipnorm=1.0,  # Gradient clipping.
    )
    # Exclude layernorm and bias terms from weight decay.
    optimizer.exclude_from_weight_decay(var_names=["bias"])
    optimizer.exclude_from_weight_decay(var_names=["gamma"])
    optimizer.exclude_from_weight_decay(var_names=["beta"])

    loss = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
    return optimizer, loss


class GPUMemoryCallback(keras.callbacks.Callback):
    def __init__(self, target_batches, **kwargs):
        super().__init__(**kwargs)
        self.target_batches = target_batches
        self.memory_usage = []

    def _compute_memory_usage(self):
        if keras.backend.backend() == "tensorflow":
            import tensorflow as tf

            memory_stats = tf.config.experimental.get_memory_info("GPU:0")
        else:
            memory_stats = {"peak": 0.0}
        # Convert bytes to GB and store in list.
        peak_usage = round(memory_stats["peak"] / (2**30), 3)
        self.memory_usage.append(peak_usage)

    def on_epoch_begin(self, epoch, logs=None):
        self._compute_memory_usage()

    def on_train_batch_begin(self, batch, logs=None):
        if batch in self.target_batches:
            self._compute_memory_usage()

    def on_epoch_end(self, epoch, logs=None):
        self._compute_memory_usage()


if __name__ == "__main__":
    EPOCHS = 2
    keras.config.disable_traceback_filtering()
    keras.mixed_precision.set_global_policy("mixed_bfloat16")

    args = get_args()
    if args.model == "gemma_2b_en":
        kagglehub.login()

    # Setup dataset
    data = []
    with open("databricks-dolly-15k.jsonl") as file:
        for line in file:
            features = json.loads(line)
            # Filter out examples with context, to keep it simple.
            if features["context"]:
                continue
            # Format the entire example as a single string.
            template = "Instruction:\n{instruction}\n\nResponse:\n{response}"
            data.append(template.format(**features))

    # Only use 16000 training examples, to keep it fast.
    data = data[:16000]

    if args.model == "gemma_2b_en":
        preprocessor = keras_nlp.models.GemmaCausalLMPreprocessor.from_preset(
            args.model, sequence_length=128
        )
        model = keras_nlp.models.GemmaCausalLM.from_preset(
            args.model, preprocessor=preprocessor
        )
        model.backbone.token_embedding.trainable = False
    elif "gpt2" in args.model:
        preprocessor = keras_nlp.models.GPT2CausalLMPreprocessor.from_preset(
            args.model, sequence_length=128
        )
        model = keras_nlp.models.GPT2CausalLM.from_preset(
            args.model, preprocessor=preprocessor
        )
        model.backbone.token_embedding.trainable = False
        model.backbone.position_embedding.trainable = False
    if args.fp8:
        model.quantize("float8")

    model.summary()
    optimizer, loss = get_optimizer_and_loss()
    model.compile(optimizer=optimizer, loss=loss, weighted_metrics=["accuracy"])
    callbacks = [
        GPUMemoryCallback(target_batches=[5, 10, 25, 50, 100, 150, 200, 300])
    ]
    model.fit(
        data, batch_size=args.batch_size, epochs=EPOCHS, callbacks=callbacks
    )
    if keras.backend.backend() == "tensorflow":
        model_memory_usage = callbacks[0].memory_usage
        print(f"GPU Memory Usage (in GB): {max(model_memory_usage)}")

    if args.fp8:
        from keras import layers

        count = 0
        for layer in model._flatten_layers(False, True):
            list_of_sublayers = list(layer._flatten_layers())
            if len(list_of_sublayers) == 1:  # leaves of the model
                if isinstance(layer, (layers.Dense, layers.EinsumDense)):
                    print(
                        layer.outputs_grad_scale.path,
                        layer.outputs_grad_scale.value,
                    )
                    count += 1
                if count > 10:
                    break

```

</details>

## Issues

- The improvements in training speed are relatively small if the model is not sufficiently large. The recommendation is to train with a model containing parameters >5B.
- You will need hardware that supports fp8 tensor cores to gain the speedups.
- It is challenging to configure for Torch without low-level operations at this time. You can check the following links for float8 training in Torch instead:
  - https://github.com/pytorch-labs/float8_experimental
  - https://github.com/NVIDIA/TransformerEngine?tab=readme-ov-file#pytorch

## References for Implementation

- https://github.com/NVIDIA/atex/tree/experimental-xla-fp8-layers/experimental/xla-fp8-layers
- https://github.com/google/flax/blob/main/flax/linen/fp8_ops.py

cc @kaixih

### CODE REVIEW

File: keras/layers/core/dense.py [hunk 1]
```diff
@@ -94,17 +97,23 @@ def __init__(
         self.kernel_constraint = constraints.get(kernel_constraint)
         self.bias_constraint = constraints.get(bias_constraint)
         self.lora_rank = lora_rank
+        self.amax_history_length = amax_history_length
```
Comment (fchollet) [inline keras/layers/core/dense.py:100]: I think this should not be set on the layer as a constructor arg, but on the DTypePolicy. If you want to configure it, call the layer with `layer.quantize(Float8DtypePolicy(amax_history_length=32))` or something like that. We may want to expose dedicate DTypePolicy subclasses to expose such arguments (this is the first one we encounter but there may be more of them in the future).

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! 👍

File: keras/backend/common/variables.py [hunk 2]
```diff
@@ -266,6 +270,20 @@ def trainable(self):
     def trainable(self, value):
         self._trainable = value
 
+    @property
+    def overwrite_with_gradient(self):
+        """Whether this variable should be overwritten by the gradient."""
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras/backend/common/variables.py]: Added.

```python
    @property
    def overwrite_with_gradient(self):
        """Whether this variable should be overwritten by the gradient.

        This property is designed for a special case where we want to overwrite
        the variable directly with its computed gradient. For example, in float8
        training, new `scale` and `amax_history` are computed as gradients, and
        we want to overwrite them directly instead of following the typical
        procedure such as gradient descent with a learning rate, gradient
        clipping and weight decaying.
        """
```

File: keras/dtype_policies/dtype_policy.py [hunk 3]
```diff
@@ -205,22 +207,22 @@ def __init__(self, name):
     def _parse_name(self, name):
         error_msg = (
             f"Cannot convert '{name}' to a QuantizedDTypePolicy. "
-            "Valid policies include "
-            "'int8_from_float32', 'int8_from_float16', 'int8_from_bfloat16', "
-            "'int8_from_mixed_float16', 'int8_from_mixed_bfloat16'."
+            f"Valid policies include {self._get_all_valid_policies()}."
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras/dtype_policies/dtype_policy.py]: Fixed

File: keras/layers/core/dense.py [hunk 4]
```diff
@@ -81,6 +83,7 @@ def __init__(
         kernel_constraint=None,
         bias_constraint=None,
         lora_rank=None,
+        amax_history_length=1024,
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras/layers/core/dense.py]: Added.

```python
        amax_history_length: Optional integer. The length of the amax history
            window used for scaling factor computation in float8 training. This
            parameter will have no effect if not in a float8 training scheme.
```

File: keras/layers/core/dense.py (hunk 1 above)
Comment (james77777778) [reply keras/layers/core/dense.py:103]: I struggled to set args to DTypePolicy and I'm uncertain if this is a good approach.

In the current codebase, configuring DTypePolicy for individual layers is challenging (as shown in #19381 ).
1. DTypePolicy is treated as a global state. Making it difficult to set different `amax_history_length` for different layers. (We also need to consider scenarios involving subclasses)
2. The serialization/deserialization may be an issue.

Based on these concerns, I adopted the approach simiar to `lora_rank` that `amax_history_length` defaults to `None` and is overridden as `1024` in `self._float8_build` if not specified.

What do you think?

Comment (fchollet) [reply keras/layers/core/dense.py:103]: I would strongly prefer to keep dtype configuration and layer constructor args separate. All dtype configuration should go through DTypePolicy.


> We also need to consider scenarios involving subclasses

If the dtype policy is a settable attribute, this is not a big issue -- you can instantiate the subclass, and then set the dtype policy on a specific layer owned by the subclass.

For serialization, the dtype policy (and its args) should be part of the layer config, under the `dtype` entry in the config dict.

`lora_rank` is simply a shortcut to 1. instantiate layer, 2. call `enable_lora()`. It's not necessary from a configuration standpoint.

Comment (james77777778) [reply keras/layers/core/dense.py]: Got it. The idea of keeping the configuration minimal for layers is solid.

As we discussed above, I have introduced a `QuantizedFloat8DTypePolicy` in this PR. I think it would clearer to use "Quantized" in the naming, as it involves quantizing and dequantizing.

BTW, I'd like to point it out that setting the dtype policy after the instantiation would not trigger the quantization process.
We need to either pass the correct `dtype_policy` (such as "float8_from_float32") though `dtype` argment in `__init__` or utilize `quantize` api.

Comment (fchollet) [reply keras/layers/core/dense.py]: > BTW, I'd like to point it out that setting the dtype policy after the instantiation would not trigger the quantization process.

Do you think it would be a good idea to implement a setter that checks whether the policy is a quantized policy, and calls `.quantize()` if it is?

File: keras/layers/layer.py [hunk 5]
```diff
@@ -668,6 +668,17 @@ def set_weights(self, weights):
                 )
             variable.assign(value)
 
+    @property
+    def dtype_policy(self):
+        return self._dtype_policy
+
+    @dtype_policy.setter
+    def dtype_policy(self, value):
```
Comment (james77777778) [inline keras/layers/layer.py:676]: I have added the setter of `dtype_policy` and it works well.

File: keras/layers/core/dense.py (hunk 1 above)
Comment (james77777778) [reply keras/layers/core/dense.py]: > Do you think it would be a good idea to implement a setter that checks whether the policy is a quantized policy, and calls `.quantize()` if it is?

This sounds like a great idea!
I think this change will also improve the UX of serializing and deserializing quantizable layers.

**Comment by james77777778** [comment]
After adding `dtype_policy` setter, we can now implement a cleaner solution to support the quantization of subclasses.

```python
import tempfile

import numpy as np

import keras
from keras import dtype_policies


@keras.saving.register_keras_serializable("MyPackage")
class MySubclass(keras.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.layer1 = keras.layers.Dense(4)
        self.layer2 = keras.layers.BatchNormalization(axis=-1)
        self.layer3 = keras.layers.ReLU()

    def build(self, input_shape):
        self._input_shape = input_shape
        self.layer1.build(input_shape)
        input_shape = list(input_shape)
        input_shape[-1] = 4
        self.layer2.build(input_shape)
        self.layer3.build(input_shape)

    def call(self, inputs):
        x = self.layer1(inputs)
        x = self.layer2(x)
        return self.layer3(x)

    def get_config(self):
        config = super().get_config()
        quantized_dtype_policies = {}
        for idx, layer in enumerate(
            self._flatten_layers(include_self=False, recursive=False)
        ):
            if isinstance(
                layer.dtype_policy, dtype_policies.QuantizedDTypePolicy
            ):
                quantized_dtype_policies[str(idx)] = layer.dtype_policy
        if quantized_dtype_policies:
            config.update(
                {"quantized_dtype_policies": quantized_dtype_policies}
            )
        return config

    @classmethod
    def from_config(cls, config):
        quantized_dtype_policies = config.pop("quantized_dtype_policies", None)
        obj = super(MySubclass, cls).from_config(config)
        if quantized_dtype_policies:
            for idx, layer in enumerate(
                obj._flatten_layers(include_self=False, recursive=False)
            ):
                if str(idx) in quantized_dtype_policies:
                    layer.dtype_policy = quantized_dtype_policies[str(idx)]
        return obj


model = keras.Sequential([keras.layers.Input([8]), MySubclass()])
model.quantize("int8")
config = model.get_config()

with tempfile.TemporaryDirectory() as tempdir:
    path = f"{tempdir}/model.keras"
    model.save(path)
    reloaded_model = keras.saving.load_model(path)  # <= failed

x = keras.random.uniform([1, 8])
y1 = model(x, training=False)
y2 = reloaded_model(x, training=False)
np.testing.assert_allclose(y1, y2)

```

**Review by fchollet — APPROVED** [review]
Excellent work -- the changes look good!
